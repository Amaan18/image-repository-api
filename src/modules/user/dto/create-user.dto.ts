import {
    IsEmail,
    IsNotEmpty,
    MinLength,
    IsAlphanumeric,
    IsArray,
    IsOptional,
} from 'class-validator';

export class CreateUserDto {
    @IsAlphanumeric()
    @IsNotEmpty()
    name: string;

    @IsEmail()
    @IsNotEmpty()
    email: string;

    @IsNotEmpty()
    @MinLength(6)
    password: string;

    @IsOptional()
    @IsArray()
    roles?: string[];
}
