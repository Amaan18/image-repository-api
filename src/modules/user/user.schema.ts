import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { Document } from 'mongoose';
import validator from 'validator';

export type UserDocument = User & Document;

@Schema({ timestamps: true })
export class User {
    @Prop({ required: [true, 'Name cannot be blank'] })
    name: string;

    @Prop({
        required: [true, 'Email cannot be blank'],
        validate: validator.isEmail,
    })
    email: number;

    @Prop({
        required: [true, 'Password cannot be blank'],
        minlength: 6,
        maxlength: 255,
    })
    password: string;

    @Prop({ default: ['user'] })
    roles: string[];
}

export const UserSchema = SchemaFactory.createForClass(User);
