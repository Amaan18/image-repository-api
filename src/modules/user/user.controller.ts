import {
    Controller,
    Get,
    Post,
    Body,
    Req,
    HttpCode,
    HttpStatus,
} from '@nestjs/common';
import { Request } from 'express';
import { CreateUserDto } from './dto/create-user.dto';
import { LoginUserDto } from './dto/login-user.dto';
import { UserService } from './user.service';

@Controller('api/v1/user')
export class UserController {
    constructor(private readonly userService: UserService) {}

    @Post('register')
    @HttpCode(HttpStatus.CREATED)
    async register(@Body() createUserDto: CreateUserDto) {
        return await this.userService.register(createUserDto);
    }

    @Post('login')
    @HttpCode(HttpStatus.OK)
    async login(@Req() req: Request, @Body() loginUserDto: LoginUserDto) {
        return await this.userService.login(loginUserDto);
    }
}
