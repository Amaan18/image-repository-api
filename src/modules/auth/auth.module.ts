import { Module } from '@nestjs/common';
import { JwtModule } from '@nestjs/jwt';
import { MongooseModule } from '@nestjs/mongoose';
import { PassportModule } from '@nestjs/passport';
import { ConfigModule } from '../config/config.module';
import { ConfigService } from '../config/config.service';
import { UserSchema } from '../user/user.schema';
import { AuthService } from './auth.service';
import { JwtStrategy } from './jwt.strategy';

@Module({
    imports: [
        ConfigModule,
        PassportModule.register({ defaultStrategy: 'jwt' }),
        MongooseModule.forFeature([{ name: 'User', schema: UserSchema }]),
        JwtModule.registerAsync({
            imports: [ConfigModule],
            useFactory: async (configService: ConfigService) => {
                return {
                    secret: configService.get('JWT_SECRET_KEY'),
                    signOptions: {
                        ...(configService.get('JWT_EXPIRATION_TIME')
                            ? {
                                  expiresIn: configService.get(
                                      'JWT_EXPIRATION_TIME',
                                  ),
                              }
                            : {}),
                    },
                };
            },
            inject: [ConfigService],
        }),
    ],
    providers: [AuthService, JwtStrategy],
    exports: [AuthService, JwtModule],
})
export class AuthModule {}
