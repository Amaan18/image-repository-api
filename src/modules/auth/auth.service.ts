import { Injectable, UnauthorizedException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { InjectModel } from '@nestjs/mongoose';
import { Model } from 'mongoose';
import { UserDocument } from '../user/user.schema';
import { JwtPayload } from './payload-interface';

@Injectable()
export class AuthService {
    constructor(
        private readonly jwtService: JwtService,
        @InjectModel('User') private readonly userModel: Model<UserDocument>,
    ) {}

    async createAccessToken(user: UserDocument) {
        const payload = { email: user.email, userId: user._id };
        return this.jwtService.sign(payload);
    }

    async validate(payload: JwtPayload) {
        const user = await this.userModel.findOne({
            _id: payload.userId,
        });
        if (!user) {
            throw new UnauthorizedException('User not found');
        }
        return user;
    }
}
