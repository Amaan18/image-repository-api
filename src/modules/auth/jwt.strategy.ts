import { ExtractJwt, Strategy } from 'passport-jwt';
import { PassportStrategy } from '@nestjs/passport';
import { Injectable } from '@nestjs/common';

import { ConfigService } from '../config/config.service';
import { UserService } from '../user/user.service';

@Injectable()
export class JwtStrategy extends PassportStrategy(Strategy) {}
