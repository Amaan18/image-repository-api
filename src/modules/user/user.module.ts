import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';
import { AuthModule } from '../auth/auth.module';
import { UserController } from './user.controller';
import { UserDocument, UserSchema } from './user.schema';
import { UserService } from './user.service';
import * as bcrypt from 'bcryptjs';

@Module({
    imports: [
        MongooseModule.forFeatureAsync([
            {
                name: 'User',
                useFactory: () => {
                    const schema = UserSchema;
                    schema.pre<UserDocument>('save', function (next) {
                        try {
                            if (!this.isModified('password')) {
                                return next();
                            }
                            const hashed = bcrypt.hashSync(this.password, 10);
                            this.password = hashed;
                        } catch (error) {
                            return next(error);
                        }
                    });
                    return schema;
                },
            },
        ]),
        AuthModule,
    ],
    controllers: [UserController],
    providers: [UserService],
})
export class UserModule {}
