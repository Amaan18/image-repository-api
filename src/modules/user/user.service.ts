import {
    BadRequestException,
    Injectable,
    NotFoundException,
} from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model } from 'mongoose';
import { AppRoles } from '../app/app.roles';
import { CreateUserDto } from './dto/create-user.dto';
import { UserDocument } from './user.schema';
import bcrypt from 'bcryptjs';
import { LoginUserDto } from './dto/login-user.dto';
import { AuthService } from '../auth/auth.service';

@Injectable()
export class UserService {
    constructor(
        @InjectModel('User') private readonly userModel: Model<UserDocument>,
        private readonly authService: AuthService,
    ) {}

    get(id: string): Promise<UserDocument> {
        return this.userModel.findById(id).exec();
    }

    getByEmail(email: string): Promise<UserDocument> {
        return this.userModel.findOne({ email }).exec();
    }

    async register(createUserDto: CreateUserDto) {
        const { email, name, password, roles: receivedRoles } = createUserDto;
        const user = new this.userModel({
            email,
            name,
            password,
            roles: receivedRoles ? receivedRoles : [AppRoles.USER],
        });
        await this.isEmailUnique(user.email);
        await user.save();
        return this.generateResponse(user);
    }

    async login(loginUserDto: LoginUserDto) {
        const user = await this.findUserByEmail(loginUserDto.email);
        await this.checkPassword(loginUserDto.password, user);
        return this.generateResponse(user);
    }

    private async isEmailUnique(email: string): Promise<void> {
        const user = await this.userModel.findOne({ email, verified: true });
        if (user) {
            throw new BadRequestException('Email most be unique.');
        }
    }

    private async findUserByEmail(email: string): Promise<UserDocument> {
        const user = await this.userModel.findOne({ email, verified: true });
        if (!user) {
            throw new NotFoundException('Wrong email or password.');
        }
        return user;
    }

    private async checkPassword(
        password: string,
        user: UserDocument,
    ): Promise<boolean> {
        const match = await bcrypt.compare(password, user.password);
        if (!match) {
            throw new NotFoundException('Wrong email or password.');
        }
        return true;
    }

    private async generateResponse(user: UserDocument) {
        return {
            name: user.name,
            email: user.email,
            accessToken: await this.authService.createAccessToken(user),
        };
    }

    private async delete(
        email: string,
    ): Promise<{
        message: string;
    }> {
        const profile = await this.userModel.deleteOne({ email });
        if (profile.deletedCount === 1) {
            return { message: `Deleted user with email: ${email}` };
        } else {
            throw new BadRequestException(
                `Failed to delete user with email: ${email}.`,
            );
        }
    }
}
