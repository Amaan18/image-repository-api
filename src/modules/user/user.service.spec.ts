import { BadRequestException, NotFoundException } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';
import { Test, TestingModule } from '@nestjs/testing';
import { Model } from 'mongoose';
import {
    closeInMongodConnection,
    rootMongooseTestModule,
} from '../../utils/MongoMemoryServer';
import { AuthModule } from '../auth/auth.module';
import { CreateUserDto } from './dto/create-user.dto';
import { LoginUserDto } from './dto/login-user.dto';
import { UserDocument, UserSchema } from './user.schema';
import { UserService } from './user.service';

describe('UserService', () => {
    let service: UserService;
    let userModel: Model<UserDocument>;

    beforeEach(async () => {
        const module: TestingModule = await Test.createTestingModule({
            imports: [
                rootMongooseTestModule(),
                MongooseModule.forFeature([
                    {
                        name: 'User',
                        schema: UserSchema,
                    },
                ]),
                AuthModule,
            ],
            providers: [UserService],
        }).compile();

        service = module.get<UserService>(UserService);
        userModel = module.get<Model<UserDocument>>('UserModel');
    });

    afterEach(async () => {
        await userModel.deleteMany({});
    });

    afterAll(async () => {
        await closeInMongodConnection();
    });

    it('should be defined', () => {
        expect(service).toBeDefined();
    });

    /*
        REGISTER TESTS
    */

    const mockCreateUserNoRole: CreateUserDto = {
        name: 'test',
        email: 'test@gmail.com',
        password: 'test123',
    };

    const mockCreateUserAdmin: CreateUserDto = {
        name: 'test',
        email: 'test@gmail.com',
        password: 'test123',
        roles: ['ADMIN'],
    };

    const mockCreateUserError: CreateUserDto = {
        name: 'test',
        email: 'test.com',
        password: 'test123',
        roles: ['ADMIN'],
    };

    it('should register user with admin role', async () => {
        const response = await service.register(mockCreateUserAdmin);
        const users = await userModel.find({});
        expect(users.length).toEqual(1);
        expect(users[0].name).toEqual('test');
        expect(users[0].roles).toContain('ADMIN');
        expect(users[0].roles).toHaveLength(1);
        expect(response.accessToken).toBeDefined();
    });

    it('should register user w/o roles', async () => {
        const response = await service.register(mockCreateUserNoRole);
        const users = await userModel.find({});
        expect(users.length).toEqual(1);
        expect(users[0].name).toEqual('test');
        expect(users[0].roles).toContain('USER');
        expect(users[0].roles).toHaveLength(1);
        expect(response.accessToken).toBeDefined();
    });

    it('should not register user if validation error', async () => {
        expect(
            service.register(mockCreateUserError),
        ).rejects.toThrowErrorMatchingSnapshot();
        const users = await userModel.find({});
        expect(users.length).toEqual(0);
    });

    it('should not register user if already exists', async () => {
        await service.register(mockCreateUserAdmin);
        expect(service.register(mockCreateUserAdmin)).rejects.toThrowError(
            new BadRequestException('Email must be unique.'),
        );
        const users = await userModel.find({});
        expect(users.length).toEqual(1);
    });

    /*
        LOGIN TESTS
    */

    const mockLoginUserValid: LoginUserDto = {
        email: 'test@gmail.com',
        password: 'test123',
    };

    const mockLoginUserInvalid: LoginUserDto = {
        email: 'invalid@gmail.com',
        password: 'test123',
    };

    it('should login user with correct creds', async () => {
        await service.register(mockCreateUserNoRole);
        const response = await service.login(mockLoginUserValid);
        expect(response.accessToken).toBeDefined();
    });

    it('should reject user with incorrect creds', async () => {
        await service.register(mockCreateUserNoRole);
        expect(service.login(mockLoginUserInvalid)).rejects.toThrowError(
            new NotFoundException('Wrong email or password.'),
        );
    });

    /*
        GET TESTS
    */
    it('should fetch user with id', async () => {
        const res = await service.register(mockCreateUserNoRole);
        const response = await service.get(res.id);
        expect(response).toHaveProperty('name');
        expect(response).toHaveProperty('id');
    });

    it('should fetch user with email', async () => {
        const res = await service.register(mockCreateUserNoRole);
        const response = await service.get(res.id);
        expect(response).toHaveProperty('name');
        expect(response).toHaveProperty('id');
    });
});
