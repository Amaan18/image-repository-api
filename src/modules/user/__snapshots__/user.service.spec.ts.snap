// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserService should not register user if validation error 1`] = `"User validation failed: email: Validator failed for path \`email\` with value \`test.com\`"`;
